libshaderc/src/shaderc_cpp_test.cc

```diff
@@ -504,6 +505,106 @@ TEST_F(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
+    std::string path;
+    std::string content;
+  };
+  // Use hashmap as a fake file system to store fake files to be included.
+  using FakeFS = std::unordered_map<std::string, FakeFile>;
+
+  TestIncluder(FakeFS& fake_fs) : fake_fs_(fake_fs){};
+
+  // Get path and content from the fake file system.
+  shaderc_includer_response* GetIncluderResponse(
+      const char* filename) override {
+    FakeFile& file_to_include = fake_fs_[filename];
+    response_.path = file_to_include.path.c_str();
+    response_.path_length = file_to_include.path.size();
+    response_.content = file_to_include.content.c_str();
+    response_.content_length = file_to_include.content.size();
+    return &response_;
+  };
+
+  // Response data is owned as private property, no need to release explicitly.
+  void ReleaseIncluderResponse(shaderc_includer_response* data) override {
+    (void)data;
+  };
+
+ private:
+  FakeFS& fake_fs_;
+  // Includer response data is stored as private property.
+  shaderc_includer_response response_;
+};
+
+TEST_F(CppInterface, Includer) {
+  const std::string kMainIncludingShader =
+      "void foo() {}\n"
+      "#include \"file_0\"\n"
+      "#include \"file_1\"\n";
+
+  TestIncluder::FakeFS fake_fs(
+      {{"file_0", {"path/to/file_0", "void bar() {}\n"}},
+       {"file_1", {"path/to/file_1", "void main() {foo(); bar();}\n"}}});
+
+  options_.SetIncluder(
+      std::unique_ptr<TestIncluder>(new TestIncluder(fake_fs)));
+
+  // Expects the compilation to succeed.
+  EXPECT_TRUE(CompilesToValidSpv(kMainIncludingShader,
+                                 shaderc_glsl_vertex_shader, options_));
+
+  // Sets compiler to preprocessing only mode, so we can check the result after
+  // preprocessing. Expect to see the content of the fake file object in the
+  // output of preprocessing.
+  options_.SetPreprocessingOnlyMode();
+  std::string preprocessing_output = CompilationOutput(
+      kMainIncludingShader, shaderc_glsl_vertex_shader, options_);
+  EXPECT_THAT(preprocessing_output, HasSubstr("#line 0 \"path/to/file_0\"\n"
+                                              " void bar(){ }\n"
+                                              "#line 2"));
+  EXPECT_THAT(preprocessing_output, HasSubstr("#line 0 \"path/to/file_1\"\n"
+                                              " void main(){ foo();bar();}\n"
+                                              "#line 3"));
+}
+
+TEST_F(CppInterface, IncluderNestedInclude) {
+  const std::string kMainIncludingShader =
+      "void foo() {}\n"
+      "#include \"file_0\"\n";
+
+  TestIncluder::FakeFS fake_fs({
+      {"file_0",
+       {"path/to/file_0",
+        "#include \"file_1\"\n"
+        "void main() {foo(); bar();}\n"}},
+      {"file_1", {"path/to/file_1", "void bar() {}\n"}},
+  });
+
+  options_.SetIncluder(
+      std::unique_ptr<TestIncluder>(new TestIncluder(fake_fs)));
+
+  // Expect the compilation to succeed.
+  EXPECT_TRUE(CompilesToValidSpv(kMainIncludingShader,
```

Comment:
Compiling all the way down to object code expands the scope of the test too far.  Should only look at the preprocessed output.